### PR TITLE
Split freshness implementation, descriptions

### PR DESCRIPTION
**Pull request by dpeng817**
The shared freshness implementation felt very hard to parse. Even though there was some shared code, it was largely a series of if-else statements, and bugs were slipping through due to obfuscation from this.

Instead, outside of shared utility functions, switch to having separate implementations for the two freshness builder methods.

Open to potentially consolidating in the future, but felt like we were currently in the worst of both worlds.

Along the way, I also switched up the descriptions and metadata to more accurately affect the freshness definition being addressed.
The new metadata fields are as follows:
- `dagster/last_cron_tick_timestamp`: if using cron-based freshness, the last completed tick of the cron.
- `dagster/lower_bound_timestamp`: the earliest that we expect the latest record to have arrived in order for the asset to be considered fresh.
- got rid of `dagster/expected_by`; we felt it was confusing to think of the failure case in terms of a deadline.
I could be convinced to make the time partition freshness descriptions a little less verbose.

There are a few associated test changes as a result of changes to description, errors, and metadata.

### CODE REVIEW

**Comment by dpeng817** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21320?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#21320** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21320?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#21295** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21295?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21251** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21251?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21247** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21247?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @dpeng817 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py [hunk 1]
```diff
@@ -100,10 +134,166 @@ def build_last_update_freshness_checks(
...
+    lower_bound_delta: datetime.timedelta,
+) -> AssetChecksDefinition:
+    params_metadata: dict[str, Any] = {TIMEZONE_PARAM_KEY: timezone}
+    if deadline_cron:
+        params_metadata[DEADLINE_CRON_PARAM_KEY] = deadline_cron
+    if lower_bound_delta:
+        params_metadata[LOWER_BOUND_DELTA_PARAM_KEY] = lower_bound_delta.total_seconds()
+
+    @multi_asset_check(
+        specs=[
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py:166]: The params here are identical across both the builder methods, right? Could we have a helper function that constructs these?

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py [hunk 2]
```diff
@@ -100,10 +134,166 @@ def build_last_update_freshness_checks(
...
+            if latest_completed_cron_tick:
+                metadata[LATEST_CRON_TICK_METADATA_KEY] = TimestampMetadataValue(
+                    latest_completed_cron_tick.timestamp()
+                )
+            if passed:
+                # If the asset is fresh, we can potentially determine when it has the possibility of becoming stale again.
+                # In the case of a deadline cron, this is the next cron tick after the current time.
+                # In the case of just a lower_bound_delta, this is the last update time plus the
+                # lower_bound_delta.
+                fresh_until = (
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py:223]: The metadata contruction logic is quite similar across the two implementations, right? Would we be able to have a helper method that's shared by them?

File: python_modules/dagster/dagster_tests/definitions_tests/freshness_checks_tests/test_last_update_freshness.py [hunk 3]
```diff
@@ -197,7 +196,7 @@ def my_asset():
             [check],
             AssetCheckSeverity.WARN,
             False,
-            description_match="Asset is overdue. Expected an update within the last 10 minutes.",
```
Comment (sryza) [inline python_modules/dagster/dagster_tests/definitions_tests/freshness_checks_tests/test_last_update_freshness.py:200]: Why remove this context? Is it inaccurate?

Comment (dpeng817) [reply python_modules/dagster/dagster_tests/definitions_tests/freshness_checks_tests/test_last_update_freshness.py:200]: It's just really verbose for a test, I feel. The description is still actually this verbose, we're just not checking the whole thing.

Comment (sryza) [reply python_modules/dagster/dagster_tests/definitions_tests/freshness_checks_tests/test_last_update_freshness.py:200]: Ahh missed that. 👍

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py [hunk 4]
```diff
@@ -100,10 +134,166 @@ def build_last_update_freshness_checks(
...
+    elif out_of_date_observation:
+        return (
+            "Asset is in an unknown state. No run has observed the asset recently enough "
+            f"(within the last {allowed_delta}) for us to know if it is fresh or overdue. Because of "
+            "this ambiguity, we marked the check as failed."
+        )
+    else:
+        return (
+            f"Asset is overdue for an update. The latest {record_type} was {actual_delta} ago, "
+            f"which is past the allowed time range of {allowed_delta}."
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py:298]: Nitpick: we typically use "range" to refer to a tuple of (start, end). "delta" or "duration" would be more accurate.

**Review by sryza — CHANGES_REQUESTED** [review]

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py (hunk 1 above)
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py:166]: I kind of feel like that's overly complicated, and the type of thing that could easily deviate at some point. My philosophy on shared code is to create a helper method when you're performing a fully self-contained task that doesn't depend on the context of the caller; and I think that the parameters are the same by circumstance but the context does matter.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py (hunk 2 above)
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py:223]: I similarly think that this obfuscates implementation.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/time_partition.py [hunk 5]
```diff
@@ -84,13 +118,169 @@ def build_time_partition_freshness_checks(
...
+
+            last_update_time_lower_bound = check.not_none(
+                last_completed_time_window,
+                "Expected a valid partitioned asset with a completed time window in order to determine valid freshness window.",
+            ).end
+
+            latest_record = retrieve_latest_record(
+                instance=context.instance, asset_key=asset_key, partition_key=expected_partition_key
+            )
+            update_timestamp = get_last_updated_timestamp(latest_record, context)
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/freshness_checks/time_partition.py:196]: I'm confused about why `update_timestamp` is figuring into the partition-based freshness calculation?

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/time_partition.py [hunk 6]
```diff
@@ -84,13 +118,169 @@ def build_time_partition_freshness_checks(
...
+    def the_check(context: AssetCheckExecutionContext) -> Iterable[AssetCheckResult]:
+        for check_key in context.selected_asset_check_keys:
+            asset_key = check_key.asset_key
+            current_timestamp = pendulum.now("UTC").timestamp()
+
+            partitions_def = cast(
+                TimeWindowPartitionsDefinition,
+                context.job_def.asset_layer.asset_graph.get(asset_key).partitions_def,
+            )
+            check.invariant(
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/freshness_checks/time_partition.py:170]: Nit: might be able to make this shorter with the following?
```python
partitions_def = check.inst(
    context.job_def.asset_layer.asset_graph.get(asset_key).partitions_def,
    TimeWindowPartitionsDefinition,
)
```

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/time_partition.py [hunk 7]
```diff
@@ -84,13 +118,169 @@ def build_time_partition_freshness_checks(
...
+                    latest_record_any_partition, last_update_time_lower_bound
+                )
+                if latest_record_any_partition is not None
+                else False
+            )
+
+            if not passed and latest_record_any_partition is not None:
+                # If this asset has been updated at all before, provide the time at which that
+                # happened as additional metadata.
+                metadata[LAST_UPDATED_TIMESTAMP_METADATA_KEY] = TimestampMetadataValue(
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/freshness_checks/time_partition.py:229]: I think that including this metadata might be confusing and distracting in the partitioned case. Easy for users to get the idea that it affects the calculation, when it actually does not, right?

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py [hunk 8]
```diff
@@ -20,11 +20,12 @@
 # Top-level metadata keys
 LAST_UPDATED_TIMESTAMP_METADATA_KEY = "dagster/last_updated_timestamp"
 FRESHNESS_PARAMS_METADATA_KEY = "dagster/freshness_params"
-# When an asset is overdue, this represents the timestamp by which the asset was expected to arrive.
-EXPECTED_BY_TIMESTAMP_METADATA_KEY = "dagster/expected_by_timestamp"
+# The latest asset record should be no earlier than this timestamp.
+LOWER_BOUND_TIMESTAMP_METADATA_KEY = "dagster/lower_bound_timestamp"
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:24]: I think this should be called something like "dagster/freshness_lower_bound_timestamp" to avoid conflicts with other kinds of lower bounds.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py [hunk 9]
```diff
@@ -240,3 +238,14 @@ def unique_id_from_asset_keys(asset_keys: Sequence[AssetKey]) -> str:
     return non_secure_md5_hash_str(
         ",".join([str(asset_key) for asset_key in sorted_asset_keys]).encode()
     )[:8]
+
+
+def is_observation_out_of_date(
+    record: EventLogRecord,
+    last_update_time_lower_bound: datetime.datetime,
+) -> bool:
+    """Determines if the provided observation is out of date with respect to the last update time lower bound."""
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:247]: What does it mean for an observation to be out of date?

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py [hunk 10]
```diff
@@ -100,10 +135,169 @@ def build_last_update_freshness_checks(
...
+                LOWER_BOUND_TIMESTAMP_METADATA_KEY: TimestampMetadataValue(
+                    last_update_time_lower_bound.timestamp()
+                ),
+            }
+            if latest_completed_cron_tick:
+                metadata[LATEST_CRON_TICK_METADATA_KEY] = TimestampMetadataValue(
+                    latest_completed_cron_tick.timestamp()
+                )
+            if passed:
+                # If the asset is fresh, we can potentially determine when it has the possibility of becoming stale again.
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py:223]: Why only "potentially"?

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py [hunk 11]
```diff
@@ -100,10 +135,169 @@ def build_last_update_freshness_checks(
...
+    asset_keys: Sequence[AssetKey],
+    deadline_cron: Optional[str],
+    timezone: str,
+    severity: AssetCheckSeverity,
+    lower_bound_delta: datetime.timedelta,
+) -> AssetChecksDefinition:
+    params_metadata: dict[str, Any] = {TIMEZONE_PARAM_KEY: timezone}
+    if deadline_cron:
+        params_metadata[DEADLINE_CRON_PARAM_KEY] = deadline_cron
+    if lower_bound_delta:
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py:166]: This will always be specified, right?

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py (hunk 1 above)
Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py:170]: Re-reviewing these, I agree with you that having a common helper function would be overly complicated in the runtime metadata case. However, over here it seems pretty straightforward – the code seems pretty much identical? If it deviates in the future, we can always split it up again.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py [hunk 12]
```diff
@@ -20,11 +20,12 @@
...
+# The latest asset record should be no earlier than this timestamp.
+LOWER_BOUND_TIMESTAMP_METADATA_KEY = "dagster/lower_bound_timestamp"
 # When an asset is fresh, this represents the timestamp when the asset can become stale again.
 FRESH_UNTIL_METADATA_KEY = "dagster/fresh_until_timestamp"
-OVERDUE_SECONDS_METADATA_KEY = "dagster/overdue_seconds"
+# If this is cron-based freshness, this is the latest tick of the cron.
+LATEST_CRON_TICK_METADATA_KEY = "dagster/latest_cron_tick_timestamp"
 
 # dagster/freshness_params inner keys
 LOWER_BOUND_DELTA_PARAM_KEY = "lower_bound_delta"
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:31]: Missed this, but this name should include the unit.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/time_partition.py (hunk 7 above)
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/time_partition.py:229]: Technically it currently does affect the calculation. I'm fine with changing it to not doing so, as I agree the value add is pretty low compared to the confusion it incurs.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py (hunk 8 above)
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:24]: fine with that.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py (hunk 9 above)
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:247]: An observation is "out of date" if it's last updated timestamp is from before the start of the current time window. If that's the case, we don't know whether it still accurately reflects reality.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py (hunk 10 above)
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py:223]: I think you're right that it's definitive.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py (hunk 11 above)
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py:166]: that's right. Artifact from the old implementation.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py (hunk 1 above)
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/last_update.py:170]: the splitting again in the future eventuality is what I'd like to avoid, but I won't belabor the point. I'll put together a shared fxn.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/time_partition.py (hunk 6 above)
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/time_partition.py:170]: good callout.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py (hunk 9 above)
Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:247]: I find that framing a little confusing - we always don't know whether it affects current reality, right?  I.e. it's always possible that the asset has changed since the last time we observed it?

Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:252]: Essentially I think there's a fundamental difference between "I observed this asset and confirmed that it hasn't been updated", and "I haven't observed this asset in a full day".

Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:252]: I agree that these are qualitatively different. However, if I understand correctly, the logic here is a heuristic that attempts to capture this qualitative difference.  I'm worried that the value in drawing this distinction is lower than the user-facing complexity that's incurred by introducing an extra variable into the pass vs. fail calculation.  It means one more thing we need to explain in our docstring, one more thing we need to explain in our result description, and one more bit of metadata to include.

Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py:252]: I don't think it factors into the pass fail calculation, or the metadata at all. It just affects the description, really.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/time_partition.py [hunk 13]
```diff
@@ -84,13 +114,134 @@ def build_time_partition_freshness_checks(
...
+            )
+            passed = latest_record is not None
+
+            metadata: Dict[str, MetadataValue] = {
+                FRESHNESS_PARAMS_METADATA_KEY: JsonMetadataValue(params_metadata),
+                LATEST_CRON_TICK_METADATA_KEY: TimestampMetadataValue(deadline.timestamp()),
+            }
+
+            # We use the latest record across all partitions to give the user additional information
+            # about the asset's state. Allows us to provide the last update time of the asset even
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/freshness_checks/time_partition.py:180]: > Allows us to provide the last update time of the asset even in the failed case

Is this still accurate?

**Review by sryza — APPROVED** [review]
I left one last comment about a comment. Otherwise, ship it I say.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/time_partition.py (hunk 13 above)
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/time_partition.py]: changed to be accurate.
